app/classes/api/base.rb

```diff
@@ -42,60 +254,55 @@ def self.instantiate_subclass(params)
     raise BadAction.new(action)
   end
 
-  # :stopdoc:
   def initialize(params = {})
     self.params = params
+    self.action = params[:action]
     self.errors = []
     initializers.each { |x| instance_exec(&x) }
   end
 
   def handle_version
-    self.version = parse_float(:version)
+    self.version = parse(:float, :version)
     if version.blank?
       self.version = self.class.version
     elsif !version.match(/^\d+\.\d+$/)
-      fail BadVersion.new(version)
+      raise BadVersion.new(version)
     else
       self.version = version.to_f
     end
   end
 
   def authenticate_user
-    key_str = parse_string(:api_key)
-    key = ApiKey.find_by_key(key_str)
+    key_str = parse(:string, :api_key)
     if !key_str
       User.current = self.user = nil
-    elsif !key
-      fail BadApiKey.new(key_str)
-    elsif !key.verified
-      fail ApiKeyNotVerified.new(key)
-    elsif !key.user.verified
-      fail UserNotVerified.new(key.user)
+      User.current_location_format = :postal
     else
+      key = ApiKey.find_by_key(key_str)
+      raise BadApiKey.new(key_str)        unless key
+      raise ApiKeyNotVerified.new(key)    unless key.verified
+      raise UserNotVerified.new(key.user) unless key.user.verified
       User.current = self.user = key.user
+      User.current_location_format = :postal
```

Comment:
Duplicates l.279. But maybe that's OK for clarity?

Reply:
I went along with rubocop for a long time on this sort of thing, but I'm increasingly firm in my disagreement.  DRY is a great principle, but like any other principle, needs to be applied in the real world.  There are cases where a tiny bit of innocuous repetition helps readability.  In this case, I think it is clearer this way because I prefer to keep the two lines User.current = xxx and User.current_location_format = xxx together.  Otherwise every time I see this code I will have to verify that there is no real reason to have them separated, that the intervening code between the lines doesn't somehow take advantage of some obscure momentary difference.  Yes, I could create yet another class method in User which sets the two simultaneously.  If that works better for you, I could swallow that.  But another thing rubocop strongly encourages which drives me out of my tree is taking single simple lines of code and moving them into single-line methods in distant parts of the code.  It just makes me have to track down dozens of silly one-line methods every time I want to figure out why a method is failing.  It's a nightmare!  (But rubcop likes it because it reduces the ABC ratio or length of whatever that cursed metric is called.)  That's why I'd prefer to keep this in-line and repeated.  I believe API still passes rubocop 100% (except where I've judiciously disabled a few metrics here and there, ha!)  Although granted I haven't run rubocop in a long time...